File: hotel_reservation/lib/models/room.dart
import 'package:hotel_reservation/models/room_category.dart';

class Room {
  final int id;
  final int roomCategoryId;
  final String roomNumber;
  final String status;
  final String? createdAt;
  final String? updatedAt;

  // Relationship
  final RoomCategory? category;

  Room({
    required this.id,
    required this.roomCategoryId,
    required this.roomNumber,
    required this.status,
    this.createdAt,
    this.updatedAt,
    this.category,
  });

  bool isAvailable() => status == 'available1';
  bool isOccupied() => status == 'occupied';
  bool isMaintenance() => status == 'maintenance';

  String getStatusText() {
    switch (status) {
      case 'available':
        return 'Tersedia';
      case 'occupied':
        return 'Terisi';
      case 'maintenance':
        return 'Maintenance';
      default:
        return 'Unknown';
    }
  }

  // Helper akses data kategori
  String get roomType => category?.name ?? '-';
  double get price => category?.price ?? 0;
  String? get description => category?.description;
  String? get imageUrl => category?.imageUrl;
  int get capacity => category?.capacity ?? 2;

  factory Room.fromJson(Map<String, dynamic> json) {
    return Room(
      id: json['id'],
      roomCategoryId: json['room_category_id'],
      roomNumber: json['room_number'],
      status: json['status'],
      createdAt: json['created_at'],
      updatedAt: json['updated'],
      category: json['category'] != null
          ? RoomCategory.fromJson(json['category'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'room_category_id': roomCategoryId,
      'room_number': roomNumber,
      'status': status,
      'created_at': createdAt,
      'updated_at': updatedAt,
    };
  }
}

File: hotel-api/app/Models/Room.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Factories\HasFactory;

class Room extends Model
{
    use HasFactory;

    protected $fillable = [
        'room_category_id',
        'room_number',
        'status',
    ];

    public function category() {
        return $this->belongsTo(RoomCategory::class, 'room_category_id');
    }

    public function reservations()
    {
        return $this->hasMany(reservation::class);
    }

    public function isAvailable()
    {
        return $this->status === 'available';
    }

    // Helper untuk akses data kategori
    public function getRoomTypeAttribute()
    {
        return $this->category->name ?? '-';
    }

    public function getPriceAttribute()
    {
        return $this->category->price ?? 0;
    }

    public function getDescriptionAttribute()
    {
        return $this->category->description ?? '';
    }

    public function getImageUrlAttribute()
    {
        return $this->category->image_url ?? null;
    }

    public function getCapacityAttribute()
    {
        return $this->category->capacity ?? 2;
    }
}
